doc/url_handler.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "url_handler.pl 1"
.TH url_handler.pl 1 "January 13th, 2020" "0.1.3" "Spawn appropriate viewer for a given URL"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
url_handler.pl \- Spawn appropriate viewer for a given URL
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBurl_handler.pl\fR \fI\s-1URL\s0\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBurl_handler.pl\fR takes an \s-1URL\s0 as argument and spawns the first executable
viewer found in either \fB\f(CB$BROWSER_\fB\f(BI\s-1SCHEME\s0\fB\fR or \fB\f(CB$BROWSER\fB\fR.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.ie n .IP "\fB\fB$BROWSER_\fB\f(BI\s-1SCHEME\s0\fB\fR" 4
.el .IP "\fB\f(CB$BROWSER_\fB\f(BI\s-1SCHEME\s0\fB\fR" 4
.IX Item "$BROWSER_SCHEME"
.PP
The user's preferred utility to browse URLs of type \fI\s-1SCHEME\s0\fR. May actually
consist of a sequence of colon-separated browser commands to be tried in
order until one succeeds. If a command part contains \f(CW%s\fR, the \s-1URL\s0 is
substituted there, otherwise the browser command is simply called with the
\&\s-1URL\s0 as its last argument. %% is replaced by a single percent sign (%), and
\&\f(CW%c\fR is replaced by a colon (:).
Examples:
.ie n .IP "$BROWSER_FTP=""wget:ncftp""" 2
.el .IP "\f(CW$BROWSER_FTP\fR=``wget:ncftp''" 2
.IX Item "$BROWSER_FTP=wget:ncftp"
.PD 0
.ie n .IP "$BROWSER_GOPHER=""gopher:lynx:links""" 2
.el .IP "\f(CW$BROWSER_GOPHER\fR=``gopher:lynx:links''" 2
.IX Item "$BROWSER_GOPHER=lynx:links"
.ie n .IP "$BROWSER_MAILTO=""mutt:pine \-url""" 2
.el .IP "\f(CW$BROWSER_MAILTO\fR=``mutt:pine \-url''" 2
.IX Item "$BROWSER_MAILTO=mutt:pine -url"
.ie n .IP "$BROWSER_NEWS=""lynx""" 2
.el .IP "\f(CW$BROWSER_NEWS\fR=``lynx''" 2
.IX Item "$BROWSER_NEWS=lynx"
.ie n .IP "$BROWSER_NNTP=""lynx""" 2
.el .IP "\f(CW$BROWSER_NNTP\fR=``lynx''" 2
.IX Item "$BROWSER_NNTP=lynx"
.PD
.PP

.ie n .IP "\fB\fB$BROWSER\fB\fR" 4
.el .IP "\fB\f(CB$BROWSER\fB\fR" 4
.IX Item "$BROWSER"
.PP
The user's preferred utility to browse URLs for which there is no special
viewer defined via \fB\f(CB$BROWSER_\fB\f(BI\s-1SCHEME\s0\fB\fR. Again it may actually consist of a
sequence of colon-separated browser commands to be tried in order until one
succeeds. If a command part contains \f(CW%s\fR, the \s-1URL\s0 is substituted there,
otherwise the browser command is simply called with the \s-1URL\s0 as its last
argument. %% is replaced by a single percent sign (%), and \f(CW%c\fR is replaced
by a colon (:).
Examples:
.ie n .IP "$BROWSER=""firefox \-a firefox \-remote openURL\e(%s\e):opera:konqueror:links2 \-g:lynx:w3m""" 2
.el .IP "\f(CW$BROWSER\fR=``firefox \-a firefox \-remote openURL\e(%s\e):opera:konqueror:links2 \-g:lynx:w3m''" 2
.IX Item "$BROWSER=firefox -a firefox -remote openURL):opera:konqueror:links2 -g:lynx:w3m"
.SH "SECURITY"
.IX Header "SECURITY"
\&\fBurl_handler.pl\fR was designed to work together with \fItin\fR\|(1) which only
issues shell escaped absolute URLs thus \fBurl_handler.pl\fR does not try hard
to shell escape its input nor does it convert relative URLs into absolute
ones! If you use \fBurl_handler.pl\fR from other applications be sure to at
least shell escape its input!
.SH "AUTHOR"
.IX Header "AUTHOR"
Urs Janssen <urs@tin.org>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
<http://www.catb.org/~esr/BROWSER/>


doc/CHANGES
-- 2.6.4 release 20250309 "" --

 028) Bjarni Ingi Gislason <bjarniig@simnet.is>
      ADD. manpage fixes (debian bug #1093395, #1096177)
      FIX. tin.1, tin.5

 027) Dennis Preiser <dennis@d--p.de>
      ADD. avoid macro in snprintf() as that might be a macro too
      ADD. fallback code for the !CPP_DOES_CONCAT case
      ADD. plural forms
      ADD. prefer MIME filename*-parameter over filename-parameter
      ADD. if a MIME filename parameter is empty after path removal fallback
           to the next filename parameter if available.
      ADD. PageToggleVerbatim (,)
      ADD. validate undeclared_charset in 'M'enu
      ADD. extend verbatim_handling options
      BUG. -L/'L' did depend on sort_article_type/sort_threads_type
           and failed for some combinations
      BUG. possible memleaks
      FIX. art.c, attrib.c, charset.c, config.c, cook.c, feed.c, filter.c
           global.c, group.c, header.c, help.c, init.c, keymap.c, lang.c
           main.c, misc.c, nntplib.c, options_menu.c, page.c, post.c, prompt.c
           rfc2045.c, rfc2046.c, rfc2047.c, save.c, select.c, thread.c
           extern.h, keymap.h, tin.h, tincfg.tbl, tin.[15]

 026) Urs Janssen <urs@tin.org>
      ADD. avoid bison warning
      ADD. use ngettext(3) for plural-forms
      ADD. avoid "install -D"
      ADD. --with-url-normalization[=[uriparser,curl] (default: yes)
           check for liburiparser or libcurl for URL normalization
      ADD. if reading via NNTP Message-ID 'L'ookup in group-, thread- and
           page level now try to find the ID via NNTP if the ID is not in
           the current group
      ADD. adjust IPv6 part of URL_REGEX
      ADD. do not check for libgsasl but zlib if not using NNTP
      ADD. move setlocale in write_overview() out of the loop
      ADD. reject invalid undeclared_charset in attributes
      ADD. if cache_overview_files is set, keep raw Subject and From
           data so the cache is build with the original data instead
           of trying recompile it from our preprocessed data.
      ADD. hash Path in overview-data
      ADD. Xref- and Path-filter to filter-menu
      ADD. ToggleInfoLastLine ('i') to OptionMenu ('M')
      ADD. configure check for getsockopt(), setsockopt()
      ADD. set TCP_USER_TIMEOUT or TCP_CONNECTIONTIMEOUT and
           TCP_RXT_CONNDROPTIME if available
      BUG. mixed up error-messages
      BUG. missing checks for empty string in From
      BUG. possible memleaks
      REM. ISO2ASC=6 (CP437) table
      FIX. active.c, art.c, auth.c, charset.c, config.c, cook.c, feed.c
           filter.c, group.c, header.c, help.c, init.c, keymap.c, lang.c
           main.c, memory.c, misc.c, newsrc.c, nntplib.c, nntps.c
           options_menu.c, page.c, pgp.c, post.c, rfc1524.c, rfc2045.c, save.c
           select.c, search.c, string.c, thread.c, xref.c, extern.h, proto.h
           tin.h, tinrc.h, parsdate.y, autoconf.h[in], tincfg.tbl
           Makefile[.in], configure[.in], tin.[15]

 025) Toomas Soome <tsoome@me.com>
      ADD. updated Estonian translation
      FIX. et.po

-- 2.6.4 release 20241224 "Helmsdale" --

 024) Bjarni Ingi Gislason <bjarniig@simnet.is>
      ADD. manpage fixes (debian bug #1089489)
      FIX. tin.1

 023) Thomas E. Dickey <dickey@invisible-island.net>
      REM. bundled-intl stuff
      ADD. disallow giving both --with-socks and --with-socks5
      ADD. try also libsocks5 and libsocksd if using --with-socks5
      FIX. intl/*, aclocal.m4, configure[.in], autoconf.hin, nntplib.h
           tcurses.h, Makefile[.in], conf-tin

 022) Enrik Berkhan <enrik@tin.org>
      ADD. avoid implicit conversion
      FIX. curses.c, misc.c, nntplib.c, rfc2047.c

 021) Matt Anton <tin@lv223.org>
      ADD. updated French translation
      FIX. fr.po

 020) Dennis Preiser <dennis@d--p.de>
      ADD. ArticleInfo-page ('\'')
      ADD. always cast arg to is*()-functions to unsigned char
      ADD. avoid implicit conversion
      ADD. formatted output for -vZ and large article counts
      ADD. avoid malloc()/free() overhead in match_group_list()
      ADD. dont_break_words tinrc option, if set don't break long
           lines inside words, just on SPACE or TAB if possible.
      ADD. charset guessing via undeclared_cs_guess in attributes
      ADD. syntax check for content subtype
      ADD. handle multiple addresses in From, Reply-To, Approved
      ADD. addresses in %T are now separated by comma instead of space
      ADD. no longer use fixed length buffers in t_cmdlineopts
      ADD. limit tinrc/attributes value strings to 1022 chars
           (they were "randomly" cut off before around 1k)
      ADD. avoid fixed length buffers in t_config, t_filter_rule and
           t_fmt
      ADD. "ternary"-operator to format-strings %F, %s, %T
      ADD. max screenwidth ('<') condition to format-strings
      ADD. avoid fixed length buffer for lines on terminal when using curses
      ADD. reorder t_config to avoid padding
      REM. MAX_MSG_HEADERS limitation
      BUG. txt_help_config_search_opt_forwards,
           txt_help_config_search_opt_forwards and
           txt_help_attrib_search_opt_backwards,
           txt_help_attrib_search_opt_forwards were mixed up
      BUG. memleak when using bsd-sockets without getservbyname()
      BUG. possible memleak on quit in version.c:upgrade_prompt_quit()
      BUG. multiple addresses in From, Reply-To, Approved (allowed since
           RFC 5536) we're not handled correct.
      FIX. active.c, art.c, attrib.c, charset.c, config.c, cook.c, debug.c
           feed.c, filter.c, getline.c, group.c, header.c, help.c, inews.c
           init.c, keymap.c, lang.c, main.c, mail.c, makecfg.c, memory.c
           misc.c, nntplib.c, nntps.c, options_menu.c, page.c, pgp.c, post.c
           prompt.c, refs.c, regex.c, rfc2046.c, rfc2047.c, save.c, screen.c
           search.c, select.c, sigfile.c, signal.c, string.c, tags.c, thread.c

tools/url_handler.pl
#! /usr/bin/perl
# example of how to call an appropriate viewer
#
# URLs must start with a scheme and shell metas should be already quoted
# (tin doesn't recognize URLs without a scheme and it quotes the metas)

use strict;
use warnings;

(my $pname = $0) =~ s#^.*/##x;
die "Usage: $pname URL" if $#ARGV != 0;

# version Number
my $version = "0.1.4";

my ($method, $url, $match, @try);
if (eval { require URI;1; } == 1) {
        $url = URI->new($ARGV[0])->canonical;
} else {
        $url = $ARGV[0];
}
$method = $url;
$method =~ s|^([^:/?#]+):.*|$1|iox;
exit 0 if ($method eq $url);

# shell escape
$url =~ s#([\&\;\`\'\\\"\|\*\?\~\<\>\^\(\)\[\]\{\}\$\010\013\020\011])#\\$1#gx;

if ($ENV{"BROWSER_".uc($method)}) {
	push(@try, split(/:/x, $ENV{"BROWSER_".uc($method)}));
} else {
	if ($ENV{BROWSER}) {
		push(@try, split(/:/x, $ENV{BROWSER}));
	} else { # set some defaults
		push(@try, 'firefox -a firefox -remote');
		push(@try, 'mozilla -remote openURL\(%s\)');
		push(@try, 'opera -remote openURL\(%s\)');
		push(@try, qw(chromium 'galeon -n' 'epiphany -n' konqueror));
		push(@try, 'lynx'); # prefer lynx over links as it can handle news:-urls
		push(@try, qw('links2 -g' links w3m surf arora));
		push(@try, 'kfmclient newTab'); # has no useful return-value on error
		push(@try, 'xdg-open'); # xdg-open evaluates $BROWSER which is unset
	}
}

for my $browser (@try) {
	# ignore empty parts
	next if ($browser =~ m/^$/ox);
	# expand %s if not preceded by odd number of %
	$match = $browser =~ s/(?<!%)((?:%%)*)%s/$1$url/ogx;
	# expand %c if not preceded by odd number of %
	$browser =~ s/(?<!%)((?:%%)*)%c/$1:/ogx;
	# reduce dubble %
	$browser =~ s/%%/%/ogx;
	# append URL if no %s expansion took place
	$browser .= " ".$url if (!$match);
	# leave loop if $browser was started successful
	last if (system("$browser 2>/dev/null") == 0);
}
exit 0;

__END__

=head1 NAME

url_handler.pl - Spawn appropriate viewer for a given URL

=head1 SYNOPSIS

B<url_handler.pl> I<URL>

=head1 DESCRIPTION

B<url_handler.pl> takes an URL as argument and spawns the first executable
viewer found in either B<$BROWSER_I<SCHEME>> or B<$BROWSER>.

=head1 ENVIRONMENT

=over 4

=item B<$BROWSER_I<SCHEME>>

=back

The user's preferred utility to browse URLs of type I<SCHEME>. May actually
consist of a sequence of colon-separated browser commands to be tried in
order until one succeeds. If a command part contains %s, the URL is
substituted there, otherwise the browser command is simply called with the
URL as its last argument. %% is replaced by a single percent sign (%), and
%c is replaced by a colon (:).
Examples:

=over 2

=item $BROWSER_FTP="wget:ncftp"

=item $BROWSER_GOPHER="gopher:lynx:links"

=item $BROWSER_MAILTO="mutt:pine -url"

=item $BROWSER_NEWS="lynx"

=item $BROWSER_NNTP="lynx"

=back

Z<>

=over 4

=item B<$BROWSER>

=back

The user's preferred utility to browse URLs for which there is no special
viewer defined via B<$BROWSER_I<SCHEME>>. Again it may actually consist of a
sequence of colon-separated browser commands to be tried in order until one
succeeds. If a command part contains %s, the URL is substituted there,
otherwise the browser command is simply called with the URL as its last
argument. %% is replaced by a single percent sign (%), and %c is replaced
by a colon (:).
Examples:

=over 2

=item $BROWSER="firefox -a firefox -remote:opera:konqueror:links2 -g:lynx:w3m"

=back

=head1 SECURITY

B<url_handler.pl> was designed to work together with L<tin(1)> which only
issues shell escaped absolute URLs thus B<url_handler.pl> does not try hard
to shell escape its input nor does it convert relative URLs into absolute
ones! If you use B<url_handler.pl> from other applications be sure to at
least shell escape its input!

=head1 AUTHOR

Urs Janssen E<lt>urs@tin.orgE<gt>

=head1 SEE ALSO

L<http://www.catb.org/~esr/BROWSER/>

=cut
